nupic/research/frameworks/greedy_infomax/models/FullModel.py

```diff
@@ -0,0 +1,911 @@
+            log_f_module_list.append(log_f_list)
+            true_f_module_list.append(true_f_list)
+            # Detach x to make sure no gradients are flowing in between modules
+            x = z.detach()
+        # Lists of lists: each list has num_modules internal lists, with each
+        # internal list containing k_predictions elements
+        return log_f_module_list, true_f_module_list
+
+    def encode(self, x):
+        # Patchify inputs
+        x, n_patches_x, n_patches_y = model_utils.patchify_inputs(
+            x, self.patch_size, self.overlap
+        )
+        x = self.encoder[0](x)
+        # Compute encoded patch-level representation for each encoder block
+        for module in self.encoder[1:]:
+            # no need to detach between modules as .encode() will only be called
+            # under a torch.no_grad() scope
+            x, out = module.encode(x, n_patches_x, n_patches_y)
+        # Return patch-level representation from the last block
+        return out
+
+
+class SparseFullVisionModel(FullVisionModel):
+    """
+    A version of the above FullVisionModel that uses sparse weights and activations.
+    :param negative_samples: number of negative samples to contrast per positive sample
+    :param k_predictions: number of prediction steps to compare positive examples.
+                          For example, if k_predictions is 5 and skip_step is 1,
+                          then this module will compare z_{t} with z_{t+2}...z{t+6}.
+    :param resnet_50: If True, uses the full ResNet50 model. If False, uses the
+                      smaller Resnet34.
+    :param grayscale: This parameter should match the transform used on the dataset.
+                      This does not actively grayscale the incoming data, but rather
+                      informs the model to use either 1 or 3 channels.
+    :param patch_size: The size of patches to split each image along both the x and
+                       y dimensions.
+    :param overlap: number of pixels of overlap between neighboring patches
+    :param sparsity: a list of sparsity values, one for each ResNetEncoder
+    :param percent_on:  a list of 3 values between (0, 1) which represent the
+    percentage of units on in each block of the ResNetEncoder
+    """
+
+    def __init__(
+        self,
+        negative_samples=16,
+        k_predictions=5,
+        resnet_50=False,
+        block_dims=None,
+        num_channels=None,
+        grayscale=True,
+        patch_size=16,
+        overlap=2,
+        sparse_weights_class=SparseWeights2d,
+        sparsity=None,
+        percent_on=None,
+    ):
+        super(SparseFullVisionModel, self).__init__(
+            negative_samples=negative_samples,
+            k_predictions=k_predictions,
+            resnet_50=resnet_50,
+            grayscale=grayscale,
+            patch_size=patch_size,
+            overlap=overlap,
+        )
+        if sparsity is None:
```

Comment:
Same as above; a user can just initiate a `FullVisionModel` instead of `SparseFullVisionModel` if there's no sparsity, correct? Either assume `sparsity` is a value in (0, 1) or raise an error if it's 0. Thoughts?

Reply:
In this case, sparsity is actually a hierarchy of dicts (in order to specify the sparsity of every layer in the network), which is why it defaults to None instead of 0.